File: internal/db/pool.go
package db

import (
	"database/sql"
	"time"

	_ "github.com/go-sql-driver/mysql"

	"github.com/y-scope/metalog/internal/config"
)

// NewPool creates a configured *sql.DB connection pool from a DatabaseConfig.
func NewPool(cfg config.DatabaseConfig) (*sql.DB, error) {
	db, err := sql.Open("mysql", cfg.DSN())
	if err != nil {
		return nil, err
	}

	poolSize := cfg.PoolSize
	if poolSize == 0 {
		poolSize = 20
	}
	minIdle := cfg.PoolMinIdle
	if minIdle == 0 {
		minIdle = 5
	}

	db.SetMaxOpenConns(poolSize)
	db.SetMaxIdleConns(minIdle)
	db.SetConnMaxIdleTime(5 * time.Minute)
	db.SetConnMaxLifetime(30 * time.Minute)

	return db, nil
}

File: crates/metalog-db/src/pool.rs
use std::time::Duration;

use metalog_config::DatabaseConfig;
use sqlx::{mysql::MySqlPoolOptions, MySqlPool};

/// Creates a MySQL connection pool from a [`DatabaseConfig`].
///
/// Pool settings:
/// - `max_connections`: from config (default 5)
/// - `min_connections`: from config (default 2)
/// - `idle_timeout`: 5 minutes
/// - `max_lifetime`: 30 minutes
pub async fn new_pool(cfg: &DatabaseConfig) -> Result<MySqlPool, sqlx::Error> {
    MySqlPoolOptions::new()
        .max_connections(cfg.pool_size)
        .min_connections(cfg.pool_min_idle)
        .idle_timeout(Duration::from_secs(300))
        .max_lifetime(Duration::from_secs(1800))
        .connect(&cfg.dsn())
        .await
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pool_options_from_config() {
        let cfg = DatabaseConfig {
            host: "localhost".into(),
            database: "test".into(),
            user: "root".into(),
            password: "".into(),
            port: 3306,
            pool_size: 10,
            pool_min_idle: 3,
        };
        // Just verify DSN is valid — actual connection requires a running DB.
        let dsn = cfg.dsn();
        assert!(dsn.starts_with("mysql://"));
        assert!(dsn.contains("test"));
    }
}
